File: sdk/samples/07_advanced/python/telemetry_dashboard.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
Telemetry Dashboard - Monitor DDS performance metrics in real-time.

Initializes HDDS telemetry, creates pub/sub on a test topic, records
latency for each write/read cycle, takes periodic snapshots, and
starts a Prometheus-compatible TCP exporter.

Usage:
    python telemetry_dashboard.py

Expected output:
    --- Snapshot #1 ---
    Messages sent: 10 | received: 10
    Latency p50: 0.120 ms | p99: 0.450 ms
    ...
    Exporter running on 0.0.0.0:4242
"""

import os
import sys
import struct
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'python'))

import hdds


BATCH_SIZE: int = 10
NUM_BATCHES: int = 5
EXPORTER_PORT: int = 4242


def now_ns() -> int:
    """Monotonic timestamp in nanoseconds."""
    return time.monotonic_ns()


def print_snapshot(snap: hdds.telemetry.MetricsSnapshot, idx: int) -> None:
    """Print a formatted metrics snapshot."""
    print(f"--- Snapshot #{idx} ---")
    print(f"  Messages sent:     {snap.messages_sent}   | received: {snap.messages_received}")
    print(f"  Messages dropped:  {snap.messages_dropped}")
    print(f"  Bytes sent:        {snap.bytes_sent}")
    print(f"  Latency p50: {snap.latency_p50_ms:.3f} ms | "
          f"p99: {snap.latency_p99_ms:.3f} ms | "
          f"p999: {snap.latency_p999_ms:.3f} ms")
    print(f"  Backpressure: merge_full={snap.merge_full_count}, "
          f"would_block={snap.would_block_count}")
    print()


def main() -> int:
    print("=" * 60)
    print("HDDS Telemetry Dashboard (Python)")
    print("=" * 60)
    print()

    hdds.logging.init(hdds.LogLevel.INFO)

    # Initialize telemetry
    metrics = hdds.telemetry.init()
    print("[OK] Telemetry initialized")

    # Create participant + endpoints
    participant = hdds.Participant("TelemetryDashboard")
    writer = participant.create_writer("TelemetryTopic", qos=hdds.QoS.reliable())
    reader = participant.create_reader("TelemetryTopic", qos=hdds.QoS.reliable())
    print("[OK] Pub/Sub created on 'TelemetryTopic'")

    # Start exporter
    exporter = hdds.telemetry.start_exporter("0.0.0.0", EXPORTER_PORT)
    print(f"[OK] Exporter running on 0.0.0.0:{EXPORTER_PORT}\n")

    # Write/read cycles with latency measurement
    for batch in range(NUM_BATCHES):
        for i in range(BATCH_SIZE):
            msg_id = batch * BATCH_SIZE + i
            payload = struct.pack('<i', msg_id)

            start = now_ns()
            writer.write(payload)
            _ = reader.take()  # best-effort read back
            end = now_ns()

            metrics.record_latency(start, end)

        # Snapshot after each batch
        snap = metrics.snapshot()
        print_snapshot(snap, batch + 1)

    # Final summary
    print("=== Dashboard Summary ===")
    final_snap = metrics.snapshot()
    print(f"Total messages sent: {final_snap.messages_sent}")
    print(f"Total bytes sent:    {final_snap.bytes_sent}")
    print(f"Final p99 latency:   {final_snap.latency_p99_ms:.3f} ms")

    exporter.stop()
    print("\n=== Telemetry Dashboard Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdk/samples/07_advanced/c/telemetry_dashboard.c
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/**
 * Telemetry Dashboard - Monitor DDS performance metrics in real-time
 *
 * Demonstrates HDDS telemetry: initializes metrics, creates pub/sub,
 * records latency for each write/read cycle, takes periodic snapshots,
 * and starts a Prometheus-compatible exporter.
 *
 * Build:
 *     cd build && cmake .. && make telemetry_dashboard
 *
 * Usage:
 *     ./telemetry_dashboard
 *
 * Expected output:
 *     --- Snapshot #1 ---
 *     Messages sent:     10   | received: 10
 *     Latency p50: 0.12 ms   | p99: 0.45 ms
 *     Bytes sent: 1280
 *     ...
 *     Exporter running on 0.0.0.0:4242
 */

#include <hdds.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "generated/HelloWorld.h"

#define BATCH_SIZE    10
#define NUM_BATCHES   5
#define EXPORTER_PORT 4242

/* Get monotonic timestamp in nanoseconds */
static uint64_t now_ns(void)
{
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ULL + (uint64_t)ts.tv_nsec;
}

static void print_snapshot(const struct HddsMetricsSnapshot *snap, int idx)
{
    printf("--- Snapshot #%d ---\n", idx);
    printf("  Messages sent:     %lu   | received: %lu\n",
           (unsigned long)snap->MESSAGES_SENT,
           (unsigned long)snap->MESSAGES_RECEIVED);
    printf("  Messages dropped:  %lu\n", (unsigned long)snap->MESSAGES_DROPPED);
    printf("  Bytes sent:        %lu\n", (unsigned long)snap->BYTES_SENT);
    printf("  Latency p50: %.3f ms | p99: %.3f ms | p999: %.3f ms\n",
           snap->LATENCY_P50_NS / 1e6,
           snap->LATENCY_P99_NS / 1e6,
           snap->LATENCY_P999_NS / 1e6);
    printf("  Backpressure: merge_full=%lu, would_block=%lu\n\n",
           (unsigned long)snap->MERGE_FULL_COUNT,
           (unsigned long)snap->WOULD_BLOCK_COUNT);
}

int main(void)
{
    printf("============================================================\n");
    printf("HDDS Telemetry Dashboard\n");
    printf("============================================================\n\n");

    hdds_logging_init(HDDS_LOG_INFO);

    /* Initialize telemetry */
    struct HddsMetrics *metrics = hdds_telemetry_init();
    if (!metrics) {
        fprintf(stderr, "Failed to initialize telemetry\n");
        return 1;
    }
    printf("[OK] Telemetry initialized\n");

    /* Create participant + endpoints */
    struct HddsParticipant *participant = hdds_participant_create("TelemetryDashboard");
    if (!participant) {
        fprintf(stderr, "Failed to create participant\n");
        hdds_telemetry_release(metrics);
        return 1;
    }

    struct HddsDataWriter *writer = hdds_writer_create(participant, "TelemetryTopic");
    struct HddsDataReader *reader = hdds_reader_create(participant, "TelemetryTopic");

    if (!writer || !reader) {
        fprintf(stderr, "Failed to create endpoints\n");
        if (writer) hdds_writer_destroy(writer);
        if (reader) hdds_reader_destroy(reader);
        hdds_participant_destroy(participant);
        hdds_telemetry_release(metrics);
        return 1;
    }
    printf("[OK] Pub/Sub created on 'TelemetryTopic'\n");

    /* Start exporter */
    struct HddsTelemetryExporter *exporter =
        hdds_telemetry_start_exporter("0.0.0.0", EXPORTER_PORT);
    if (exporter) {
        printf("[OK] Exporter running on 0.0.0.0:%d\n\n", EXPORTER_PORT);
    } else {
        printf("[WARN] Exporter failed to start (continuing without)\n\n");
    }

    /* Write/read cycles with latency measurement */
    for (int batch = 0; batch < NUM_BATCHES; batch++) {
        for (int i = 0; i < BATCH_SIZE; i++) {
            HelloWorld msg = {.id = batch * BATCH_SIZE + i};
            snprintf(msg.message, sizeof(msg.message), "sample_%d", msg.id);

            uint8_t buf[256];
            size_t len = HelloWorld_serialize(&msg, buf, sizeof(buf));

            uint64_t start = now_ns();
            hdds_writer_write(writer, buf, len);

            /* Read back (best-effort, may not always get data) */
            uint8_t rbuf[256];
            size_t rlen = 0;
            hdds_reader_take(reader, rbuf, sizeof(rbuf), &rlen);
            uint64_t end = now_ns();

            hdds_telemetry_record_latency(metrics, start, end);
        }

        /* Take snapshot after each batch */
        struct HddsMetricsSnapshot snap;
        if (hdds_telemetry_snapshot(metrics, &snap) == HDDS_OK) {
            print_snapshot(&snap, batch + 1);
        }
    }

    /* Final summary */
    printf("=== Dashboard Summary ===\n");
    struct HddsMetricsSnapshot final_snap;
    if (hdds_telemetry_snapshot(metrics, &final_snap) == HDDS_OK) {
        printf("Total messages sent: %lu\n",
               (unsigned long)final_snap.MESSAGES_SENT);
        printf("Total bytes sent:    %lu\n",
               (unsigned long)final_snap.BYTES_SENT);
        printf("Final p99 latency:   %.3f ms\n\n",
               final_snap.LATENCY_P99_NS / 1e6);
    }

    /* Cleanup */
    if (exporter) hdds_telemetry_stop_exporter(exporter);
    hdds_reader_destroy(reader);
    hdds_writer_destroy(writer);
    hdds_participant_destroy(participant);
    hdds_telemetry_release(metrics);

    printf("=== Telemetry Dashboard Complete ===\n");
    return 0;
}
